### DISCUSSION TITLE
How to get type templates of extendend class/interface

### DISCUSSION THREAD

**Discussion by thomasvargiu** [vimeo/psalm#9255] [Plugin development]
Hi,

I'm in the situation where I have this function:

```php
/**
 * @template A
 * @psalm-pure
 * @param (callable(): (PromiseInterface<A>)) $function
 * @return Closure(): PromiseInterface<A>
 */
function async(callable $function): callable
{
}
```

and I need to write a plugin to be sure that the returned closure requires the same arguments of the `$function` argument.

The plugin is `FunctionReturnTypeProviderInterface` and I'm able to check whether the `$function` returns a `PromiseInterface` instance with:

```php
class AsyncHandler implements FunctionReturnTypeProviderInterface
{
    public static function getFunctionReturnType(FunctionReturnTypeProviderEvent $event): ?Type\Union
    {
        $source = $event->getStatementsSource();
        
        // [...]
        
        // check whether the $function callable returns a PromiseInterface instance
        $isPromise = $source->getCodebase()->isTypeContainedByType(
            $functionParam->return_type ?? Type::getMixed(),
            new Type\Union([Type\Atomic\TNamedObject::createFromName(PromiseInterface::class)])
        );
    }
```

Now, we can imagine we have other implementations that extends `PromiseInterface` with other templates, I want my function to always returns a `PromiseInterface<value>`, but I need to find the right template value.

If I have a type like this:
```php
/**
 * @template B
 * @template A
 * @template-implements PromiseInterface<A>
 */
class PromiseExtended implements PromiseInterface {}
```

**The question is**: how can I have the template value of `PromiseInterface` from another type that extends it?
Is it a missing available feature from public classes? I don't want to use internal classes.

**Comment by orklah** [comment]
I'm not sure I understood everything correctly, but you should be able to retrieve parameters through `$event->getCallArgs()`, then get their types with something like `$statements_source->getNodeTypeProvider()->getType($call_args[0]->value)` and the get the type of the template from the Union you retrieved I think
